### ISSUE TITLE
test-case: general, endian or amigaos4 specific #3

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#50]
After spending 2 days for porting legacy test-case opengl examples, found again one which didn't works correctly. That can be of course as usuall problems of amigaos4, or endianes, or they can be general as well.

So, test case  demonstrates using mipmaps for texture maps. To overtly show the effect of mipmaps, each mipmap reduction level has a solidly colored, contrasting texture image. Thus, the quadrilateral which is drawn is drawn with several different colors.

There is source code (amigaos4/ogles2):
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_3/mipmap_gl4es.c

By keys '+' and '-' you change the size (and colors should be changed as wel).

There is image of how it looks like on our legacy opengl1.2:
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_3/mipmap_mgl.jpg

There is image (wrong) how it looks like over gl4es:
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_3/mipmap_gl4es.jpg
Pressing on '+' and "-" change only size, but not color. Color always yellow

There is debug output from gl4es version:
http://kas1e.mikendezign.com/aos4/gl4es/tests/testcase_3/debug.txt

**Comment by ptitSeb** [comment]
Mmmm, I don't see anything wrong, exept, it's again using Texture 0...
I have to adapt the sample on my side to tes, but in the mean time, you can may be try the `glGenTextures(...)`/`glBindTexture(...)` trick, just to be sure it's not a texture 0 side-effect.

**Comment by kas1e** [comment]
Just tried as before right after makeImages() added:
	
	GLuint tex;
	glGenTextures(1, &tex);
	glBindTexture(GL_TEXTURE_2D, tex);

And that didn't help.

**Comment by ptitSeb** [comment]
Ah, ok.
will test on my side.

**Comment by ptitSeb** [comment]
Working with no issue on my side on the Pandora...
You need to ask the OGLES2 guy if Mipmap are handled... I don't see anything wrong.

**Comment by ptitSeb** [comment]
One point to note: the mipmaps are first uploaded, then the texture min filter is switched to `GL_LINEAR_MIPMAP_LINEAR`

This is not wrong, but maybe the OGLES2 driver prefer it the other way.
To test, move all 4 `glTexParameterf` before the `glTexImage2D`

**Comment by kas1e** [comment]
Tested by move all 4 glTexParameterf before the glTexImage2D : still same :(
Can it be there that we meet with Endianes ? I just remember that when endianes wrong, it usuall to have all in Yellow most of the time :)

**Comment by ptitSeb** [comment]
I don't think it's endianess yet. `GL_RGB` / `GL_UNSIGNED_BYTE` is local to the machine, wand we saw in a previous test that it was correct.
I think the level>0 are not uploading in OGLES2 driver, or in Warp3D...

**Comment by ptitSeb** [comment]
(btw, did you try some real app/game or not yet. did you also compare speed between gl4es and minigl?)

**Comment by ptitSeb** [comment]
Maybe it's just the `GL_LINEAR_MIPMAP_LINEAR` that is not supported. Can you try `GL_LINEAR_MIPMAP_NEAREST` instead?

**Comment by kas1e** [comment]
GL_LINEAR_MIPMAP_NEAREST also didn't help.. 

Will bring all that info to ogles2's author.

As for speed: GL4ES show practically the same speed as MiniGL on those small examples, which is good, as there is almost no workload , but under the hood a lot of happens. All the difference will come from real apps of course. But for that i firstly need to integrate it to our SDL1, and thinking about how to do it all properly:  https://github.com/capehill/os4sdl/commits/1.2.15

And then can port some real games/apps.

**Comment by ptitSeb** [comment]
I see.

For sdl integration, the `agl.c`/`agl.h` I wrote as part of gl4es may help. All function for Opening / Closing the Context, SwapBuffer, and even a `aglGetProcAddress(...)` as a bonus (because software may want to get functio pointer to extensions, like for shaders).

**Comment by kas1e** [comment]
Sended all info to ogles2 develope, let's see what he answer. But i checked in readme for driver : mipmaping definately should works.

As for our agl/agl.c , it bring me some errors in current form. First ones because there need to add 
#include <stdarg.h>, and then after:

src/agl/agl.c: In function 'agl_context_find':
src/agl/agl.c:50: warning: 'return' with a value, in function returning void
src/agl/agl.c: In function 'aglCreateContext':
src/agl/agl.c:81: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglCreateContextTags':
src/agl/agl.c:90: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglDestroyContext':
src/agl/agl.c:98: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglMakeCurrent':
src/agl/agl.c:106: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglSwapBuffers':
src/agl/agl.c:114: error: 'glstate' undeclared (first use in this function)
src/agl/agl.c:114: error: (Each undeclared identifier is reported only once
src/agl/agl.c:114: error: for each function it appears in.)
src/agl/agl.c:116: warning: implicit declaration of function 'flush'
src/agl/agl.c:119: warning: implicit declaration of function 'bitmap_flush'
src/agl/agl.c:121: error: 'globals4es' undeclared (first use in this function)
src/agl/agl.c:123: warning: implicit declaration of function 'unbindMainFBO'
src/agl/agl.c:124: warning: implicit declaration of function 'blitMainFBO'
src/agl/agl.c:129: error: dereferencing pointer to incomplete type
src/agl/agl.c:134: warning: implicit declaration of function 'bindMainFBO'
src/agl/agl.c: In function 'aglSetBitmap':
src/agl/agl.c:141: error: dereferencing pointer to incomplete type

**Comment by ptitSeb** [comment]
I need a .h where `ULONG` is defined, or at least just the `typedef` of it.

**Comment by kas1e** [comment]
Hm, strange, it is defined in the exec/types.h, which you include in agl.h already at begining ..

**Comment by kas1e** [comment]
Maybe it because you give functions the same names as in ogles2 includes ?

**Comment by ptitSeb** [comment]
Mmm, I don't know. I have made a commit with some attept at fixing all the error, but I'm not sure I have fixed the "dereferencing pointer" error.

**Comment by kas1e** [comment]
Yep, they still here..

**Comment by ptitSeb** [comment]
Can you copy/paste the error messages here?

**Comment by kas1e** [comment]
Sure:

src/agl/agl.c: In function 'AaglCreateContext':
src/agl/agl.c:81: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglCreateContextTags':
src/agl/agl.c:90: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglDestroyContext':
src/agl/agl.c:98: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglMakeCurrent':
src/agl/agl.c:106: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglSwapBuffers':
src/agl/agl.c:118: error: dereferencing pointer to incomplete type
src/agl/agl.c: In function 'aglSetBitmap':
src/agl/agl.c:126: error: dereferencing pointer to incomplete type

**Comment by ptitSeb** [comment]
Ah, I need to `#include <interfaces/ogles2.h>` I guess. Now that I have moved away `gl.h` from gl4es, it should works.

**Comment by kas1e** [comment]
Yeah, now it compiles fine.

Starting to try to integrate GL4ES to SDL1 from that file:
https://github.com/capehill/os4sdl/blob/1.2.15/src/video/amigaos4/SDL_os4gl.c

At moment stuck a bit, as minigl context have:
                                     MGLCC_FrontBuffer,		hidden->m_frontBuffer,
                                     MGLCC_BackBuffer,		hidden->m_backBuffer,

And all stuff (including SDL_GL_SwapBuffers), heavy based on , while for ogles we don't have those buffers. 

At moment can create ogles2 context here, link with test sdl1 + libgl4es, + create test SDL/opengl app, and it runs : black window open, and i can see debug output come from gl4es. Through, all black, as SDL_GL_SwapBuffers use those "temporary" things, and dunno at moment how to deal. Will continue tomorrow, and hope ogles2 author will answer something about main topic.

**Comment by kas1e** [comment]
That what have currently:
http://kas1e.mikendezign.com/aos4/gl4es/sdl1/SDL_os4gl.c

Through as it all about those back/front buffers probably all of this need to delete and totally rewrite

**Comment by ptitSeb** [comment]
You can probably start by ignoring all this double buffering stuff and just create the context and 
 simply `aglSwapBuffer()` and It should work.

**Comment by kas1e** [comment]
Ok, Daniel fix another issue in ogles2 driver (thanks to you and to gl4es in whole, we fix good bunch of errors with which we can meet soon or later). It was indeed manually broken setting a texture's mipmap-levels via glTexImage2D and a level > 0.  The last supplied bitmap would essentially reset the texture and kill the previously supplied tex-level,  because due to a stupid typo the lib would believe that the texture's base level size was modified.

As for integration to SDL1, doesn't you mind if i didn't close that issue for a little longer, as we just at it ?

There is what i have now (refresh need for redownload):
http://kas1e.mikendezign.com/aos4/gl4es/sdl1/SDL_os4gl.c
http://kas1e.mikendezign.com/aos4/gl4es/sdl1/SDL_os4video.h
(at end of .h you can see i add void *gl, so for create ogles2 context)

In that form, i can run opengl test example, sdl window initializes , i see that gl4es debug output starts, but nothing in the window draws. So, then i just added glFlush() before IOGLES2->aglSwapBuffers(); and it crashes :) 

Probably time to use code from your aglSwapBuffer version ..

**Comment by kas1e** [comment]
Btw, you have little typo in GL.c, forgotten ";" at end of static int amiga_old_batch

**Comment by kas1e** [comment]
Ok, seems problem is with: https://github.com/capehill/os4sdl/blob/1.2.15/src/video/amigaos4/SDL_os4_wrapper.c

One of developers says me that i need to be able to get function addresses, it's an SDL feature. You need the ProcAddress mechanism. If gl4es has it internally you may not need a wrapper table then.

So, probably i can delete whole content of that file ? But then, what to do with AmiGetGLProc() at end of file, as i have undef-reference to it if i comment all the data in whole file.

**Comment by ptitSeb** [comment]
Regarding your `SDL_os4gl.c` I suggest you `#include <GL4ES/agl.h>` and romove all `IOGLES2->` reference and simply use `aglXXXX(...)` function for gl4es directly.
Remove the `AmiGetGLProc(...)` reference also and use `aglGetProcAddress(...)` instead.

**Comment by kas1e** [comment]
Yeah ! http://kas1e.mikendezign.com/aos4/gl4es/sdl1/sdl1.jpg

Yes, i just use your aglGetProcAddress as you say.

Through, if i use just IOGLES2->aglSwapBuffer(); then nothing drawns. But if i use your aglSwapBuffer(), then it works.  Probably can just reuse your code from agl.c (it just i not mantainer of SDL1, so want to make all "easy to understand" for that one who will add later gl4es support in).

**Comment by ptitSeb** [comment]
Code will be something like:
```
/*
 * initialize OGLES2 context
 */
int os4video_GL_Init(_THIS)
{

	struct SDL_PrivateVideoData *hidden = _this->hidden;
	
	dprintf("Initializing OpenGL.. ");

		
	 hidden->IGL=aglCreateContextTags(0,
				OGLES2_CCT_WINDOW,(ULONG)hidden->win,
				OGLES2_CCT_DEPTH,16,
				OGLES2_CCT_STENCIL,8,
				OGLES2_CCT_VSYNC,0,
				OGLES2_CCT_SINGLE_GET_ERROR_MODE,1,
			TAG_DONE);


	if (hidden->IGL)
	{
		_this->gl_config.driver_loaded = 1;
	
		aglMakeCurrent(hidden->IGL);		
		
		hidden->OpenGL = TRUE;

		printf("opengl mode true , yeah !\n");
		return 0;
	}
	else
	{
		_this->gl_config.driver_loaded = 0;
		SDL_SetError("Failed to create OGLES2 context");
	}

	return -1;

}

void os4video_GL_Term(_THIS)
{

}

int	os4video_GL_GetAttribute(_THIS, SDL_GLattr attrib, int* value)
{
   return 0;
}

int	os4video_GL_MakeCurrent(_THIS)
{
	return 0;
}

void os4video_GL_SwapBuffers(_THIS)
{

	printf ("we in the SDL_GL_SwapBuffers!\n");
	aglSwapBuffers();

	
}


void *os4video_GL_GetProcAddress(_THIS, const char *proc) {
	void *func = NULL;

	if ( !_this->gl_config.driver_loaded )
	{
		if (os4video_GL_Init(_this) != 0)
		{
			return NULL;
		}
	}

	func = aglGetProcAddress(proc);
	return func;
}

int os4video_GL_LoadLibrary(_THIS, const char *path) {
	/* Library is always open */
	_this->gl_config.driver_loaded = 1;

	return 0;
}
```

**Comment by ptitSeb** [comment]
Do as think is best. `agl.c` wil probably note change a lot, as I don't have any Amiga to test it, I wont make much evolution to that file.

**Comment by kas1e** [comment]
yeah, check previous comment :)

**Comment by kas1e** [comment]
Yeah, thanks a bunch ! Now .. now should i try to port some sdl game, or, firstly need to deal with double-buffering ?:)

**Comment by ptitSeb** [comment]
No, don't bother with double buffering. I'm pretty sure it's fine as it's implemented.
Go try some sdl game. I assume you have some ready to try (I do have a couple of SDL/GL game also on my github account).

**Comment by ptitSeb** [comment]
Or see with OGLES2 doc if there is some `OGLES2_CCT_XXXX` for double buffering if you are really woried.

**Comment by kas1e** [comment]
I will start from very simple ones (as for sure ogles2 drive will bring us some surprises again:) ).

I will try now cadog_adventures, pretty fine and simple platformer , SDL1/OGL, i did port it back in past over legacy minigl, so, kind of good canditate for first test. There is souce if you want to port it to pandora as well (it mostly ./configure ; make if i remember right):

http://www.martin.st/software/cadog/

**Comment by ptitSeb** [comment]
Yeah, already ported games are a good starting point!

Didn't knew this game, I'll port to the Pandora I think :p

**Comment by kas1e** [comment]
Ok, seems still there needs to be more support in SDL for gl4es, as once i run Cadog in full screen, it says "printf("No fullscreen modes available!\n"); ", but then, when i run it in window mode "-w" , it starts, open window, but then nothing in, just white window :) while should be : http://os4depot.net/filedata/snapshot/5461_1_snap.jpg

**Comment by ptitSeb** [comment]
I don't have any idea for the fullscreen problem, you should ask a SDL1/AmigaOS4 maintenair.

For the White window, it may be some Texture/Endianness issue. You may have to add some log to `src/gl/texture.c` to see what kind of texture it's uploading.

**Comment by kas1e** [comment]
Yeah some more tests with SDL1 need it for sure. I also tried for now some tetris clone, called : LettersFall (also opengl+sdl1). That one build too, but also give black screen. Music plays, i can press on keys and heart sounds, so.. probably our SDL1 / GL4ES intergration need some love for first :)

**Comment by ptitSeb** [comment]
There must be something in the init phase or the texture loading phase that break something.

Unfortunatly, I don't know either games. I'll try Cadog tonight o the Pandora.

I was thinking of something simple, like bloboats, but I don't find a reliable source code depot. It's on Debian, so there must be some git repo somewhere. Bloboats has not a lot of textures, and drawing is mainly large block of `glBegin()`/`glEnd()`.
stuntcarremake on my github also has a simple init phase, and not a lot of texture.

The point would be to get some simple SDL+Texture loaded from a file (a PNG, using SDL_image for example) to test this kind of things works.

(on the 2 games you tried, there was no particular error on LIBGL side? did you try some gl4es debug logging to see if everything seems normal?).

**Comment by kas1e** [comment]
There is debug logging from LetterFall (3mb) , in which i run the game, wait till heard music, press randomly in window (so can heard click), and then press "exit" from :
http://kas1e.mikendezign.com/aos4/gl4es/games/lettersfall/letters_fall_debug.txt

I am almost sure it is SDL1 which need to be fixed more.. Did you see anything "bad" in logs ?

**Comment by kas1e** [comment]
I cookie up another test case (more intersting now), and it also show me black screen :) (all previous tests works ok). I will create new BZ for , few secs.

**Comment by ptitSeb** [comment]
I have seen nothing wrong on previous log.

Texture loaded seem fairly standard, with mainly GL_RGBA/GL_UNSIGNED_BYTE and some GL_BGRA/GL_UNSIGNED_BYTE (but they will be transformed into RGBA by gl4es).

If you have a simple test case, that's good.

**Comment by kas1e** [comment]
I think its just SDL. Because it also use GL_Set_Attribute probably, and that one which i remove from os4_gl.c file. Probably with simple tests cases its all ok, but once code want to use SDL  for real, then fail.

**Comment by kas1e** [comment]
I mean GL_GetAttribute, but point is that SDL should be fixed for sure.

**Comment by ptitSeb** [comment]
Most games just use `GL_Set_Attribute` but don't check back `GL_Get_Attribute`  to see if all is correct. So not sure it's that.

**Comment by ptitSeb** [comment]
I have tried cadog on the Pandora, using GLES2 backend, and it's working perfectly.

I have noticed that it first do a fade from white at the very beggining. Would it be possible it get stuck at the all white and never fade to the menu on AmigaOS for some reason?

**Comment by kas1e** [comment]
Dunno.. I think there something general in our SDL.. That private video structure may be of use in some parts of sdl..

From other side: sdlgears and sdl-draw-triangle (both opengl), works fine.

Maybe i need to start with something dead simple as you say. Through Cadog and LettersFall simple enough..

**Comment by ptitSeb** [comment]
Yeah. Maybe try to put some `printf` in the drawing loop, just to be sure it advance, and also a `printf` in `aglSwapBuffer(...)` to be sure swapbuffer is called and do something.

**Comment by kas1e** [comment]
Found sources of bloboats:
http://http.debian.net/debian/pool/main/b/bloboats/bloboats_1.0.1.dsfg.orig.tar.gz

Already compiled it with our Legacy minigl: works.

Now to compile it over SDL1-GL4ES !

**Comment by ptitSeb** [comment]
I have just tried bloboats with gl4es on an Odroid N1, and yeah, it works well with no issues.

**Comment by kas1e** [comment]
Ok, made bloboats  sdl1/gl4es version as well. Good news : IT RUns ! 

Through, as expected, wrong colors :) At least time for endian fixes come (i hope it is endianes this time ..)

There screenhot of menu from minigl version : http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/bloboads_menu_minigl.jpg

There screenhot of menu from GL4ES version : http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/bloboads_menu_gl4es.jpg

There debug log from gl4es version:
http://kas1e.mikendezign.com/aos4/gl4es/games/bloboads/debug_gl4es.txt

**Comment by ptitSeb** [comment]
Mmm, I have some color issue on the Pandora also, that I have to fix. But it's not the same as the color issue you have.
Texture is BGRA/UBYTE, it's classic, I'm not sure why there is a problem, I have to work on that...

**Comment by kas1e** [comment]
It's like in case with GL4ES, it has some "tranparency" where should be black color. 
What kind of color issues you have ?

**Comment by kas1e** [comment]
As issue fixed in our ogles driver and sdl1 seems works, we can close that issue probably and keep as opened another one where we discuss porting stuff :)
